### DISCUSSION TITLE
[Chapter 08] 도커 애플리케이션에 관련된 질문

### DISCUSSION THREAD

**Discussion by JoisFe** [cowaine/spring-boot-for-msa#87] [Chapter 08. Data Persistence]
이번 Chapter 8 에서 JPA 학습 이전에 MySQL 설치를 하는데 도커 컨테이너를 통해 MySQL을 설치하여 사용 합니다.
로컬 환경에 영향을 받지 않고 독립적으로 각 애플리케이션에 필요한 버전을 사용할 수 있어 자주사용하여 많이 사용하고 있습니다.

그런데 스터디원들에게 궁금한점이 있습니다.
런타임 컨테이너인 도커 경우 실제로 리눅스 기반의 컨테이너 기술을 사용하도록 설계되었습니다.
스터디원 분들 대부분 Windows, MacOS 환경에서 개발을 하고 계실텐데 따라서 도커 데스크톱 혹은 런쳐 데스크톱과 같은 도커 애플리케이션 툴을 사용하여 Docker를 사용하실 것으로 예상됩니다.

- [참고 자료](https://docs.docker.com/get-started/overview/)
<img width="776" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/3dcfee1b-9288-4d5d-8d0c-0e4784afe049">

```
Docker Desktop is an easy-to-install application for your Mac, Windows or Linux environment that enables you to build and share containerized applications and microservices. Docker Desktop includes the Docker daemon (dockerd), the Docker client (docker), Docker Compose, Docker Content Trust, Kubernetes, and Credential Helper. For more information, see [Docker Desktop](https://docs.docker.com/desktop/).
```

Q) Docker Desktop 경우 최근에 Enterprise 용으로 유료화가 되었습니다. 그래서 저는 현재 Rancher Desktop을 대용으로 사용 중인데, 다른 좋은 대체제가 있을까요?

Q) 저는 현재 `Rancher Desktop` 사용 중이고 
``` bash
docker run -d -p 3306:3306 -e MYSQL_ROOT_PASSWORD=1q2w3e4r \
-v $HOME/DockerData/spring-tour-mysql:/var/lib/mysql \
--name spring-tour-mysql mysql:latest \
--character-set-server=utf8mb4 \
--collation-server=utf8mb4_unicode_c
```

책에 나와있듯 해당 명령어를 입력 시
<img width="1821" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/8960dba0-c040-4d14-a2b6-e3065699debc">

위와 같은 에러가 발생합니다.

- 같은 m1 silicon chip을 사용 중인 @KIrisation 님은 `Docker Desktop` 을 사용 중인데요, 위와 같은 문제는 발생하지 않습니다.
- 해당 경로의 폴더에 권한을 주어도 같은 문제가 발생하여 확인해보니 

- [참고 자료](https://www.google.com/search?q=%08lima+linux+volume+chown&sca_esv=577395672&biw=1920&bih=995&sxsrf=AM9HkKnpyfLHKD9qJDfxPESPyQXcDMWF8Q%3A1698478142963&ei=Prg8ZbC2Osrt-AafxK-wBg&ved=0ahUKEwiwzfqSnJiCAxXKNt4KHR_iC2YQ4dUDCBA&uact=5&oq=%08lima+linux+volume+chown&gs_lp=Egxnd3Mtd2l6LXNlcnAiGAhsaW1hIGxpbnV4IHZvbHVtZSBjaG93bjIFEAAYogQyBRAAGKIEMgUQABiiBDIFEAAYogRI380QUP-fEFiYzBBwAHgCkAEAmAG8AaAB-hyqAQQwLjI1uAEDyAEA-AEBwgIEEAAYR8ICCxAAGIAEGLEDGIMBwgIKEAAYgAQYFBiHAsICCBAuGLEDGIAEwgIEEAAYA8ICCBAAGIAEGLEDwgIIEC4YgAQYsQPCAgQQIxgnwgIHECMYigUYJ8ICERAuGIAEGLEDGIMBGMcBGNEDwgIOEC4YgAQYsQMYxwEY0QPCAgcQABiABBgKwgIFEC4YgATCAgUQABiABMICCxAuGIMBGLEDGIAEwgILEC4YgAQYxwEY0QPCAgQQABgewgIGEAAYHhgKwgIGEC4YHhgKwgIGEAAYCBgewgIIEAAYCBgeGA_CAgUQIRigAcICBxAhGKABGArCAgQQIRgV4gMEGAAgQYgGAZAGCg&sclient=gws-wiz-serp)
```
I too am keen to see this resolved. Having been a Docker Desktop user, and developing on my M1 mac, I have no issues with permissions when starting docker containers. All my project files are owned by my id, however the docker containers that start (eg: Postgres or mysql) while they may attempt to chown and change permissions on files as part of their normal startup process - those attempts don’t fail and those containers run happily.

On rancher desktop, those file mod permissions fail with permission denied errors - and if I change the uid/gid (from the host) to what the process is running on in the container, then those files are not visible and file not found errors pursue.

I changed from Docker Desktop on my company issued device, as I didn’t want my company to think that I was using commercial software without meeting the license requirements, nor did I want Docker to think I was doing the same.
```

- 제가 겪은 고민과 유사한 질문을 올린 사람들이 꽤 있었으나 명확한 해답을 얻지 못했습니다.
- lima 기반 툴을 사용 중인 경우 이러한 문제가 있는 것 같은데, 혹시 원인을 아시는 분 있으면 공유해주시면 감사하겠습니다.

**Comment by corock** [comment]
> Q. Docker Desktop 경우 최근에 Enterprise 용으로 유료화가 되었습니다. 그래서 저는 현재 Rancher Desktop을 대용으로 사용 중인데, 다른 좋은 대체제가 있을까요?

Docker Desktop 의 유료화로 인한 [가격 정책](https://www.docker.com/pricing)을 확인하고, 유료화 대상에 속하는지 먼저 볼 필요가 있겠네요.
Docker Free 구독 정책이 Personal 플랜으로 변경되면서, 250명 이상의 직원이 속한 매출 1,000만 달러 이상인 경우 유료 플랜을 구독해야 합니다.
다시 말하면, 개인 사용자나 250인 이하, 그리고 1,000만 달러 미만의 매출이 나지 않는다면 무료라는 얘기가 됩니다.
그래서 위 조건에 해당하지 않는다면 저는 Docker Hub 의 최신 소식 등을 확인할 수 있는 점에서 Docker Desktop 을 계속 사용할 예정입니다.
그래도 대체제를 찾는다면 언급된 [Lima](https://github.com/lima-vm/lima) 나 Minkube, Podman 정도가 검색되는 것 같습니다.

cf. [도커 데스크톱, 대기업 사용자에게는 유료화된다](https://www.ciokorea.com/news/206529#csidx114a089318192ce8ddd024184470a9c)

> Q) 저는 현재 Rancher Desktop 사용 중이고
> (...)
> lima 기반 툴을 사용 중인 경우 이러한 문제가 있는 것 같은데, 혹시 원인을 아시는 분 있으면 공유해주시면 감사하겠습니다.

이건 제가 Rancher Desktop 을 사용해 보지 않아서 잘 모르겠네요. 😭
다만 최근에 아두이노 코딩을 해야 하는 상황이 있었는데, [Arduino Forum](https://forum.arduino.cc) 사이트를 참고하다 보면 재밌는 현상이 있습니다.
뭐만 하면 안된다 질문이 지금도 꾸준히 올라오고 있거든요. 아마 명확한 해답을 얻지 못했다면
1. 해답이 어딘가에 있는데 아직 찾지 못함
2. 대체제 자체의 이슈로 해당 개발자가 작업하여 업데이트를 기다림

정도의 방법이 있을 것 같습니다...

**Comment by Irisation23** [comment]
### 질의 응답 이외의 CheckPoint

1. 도커엔진이 유료화 되지는 않았음.
2. Docker Desktop -> Rancher Desktop 이전 중 겪었던 문제가 있었음.
  - 리눅스를 lima로 추상화 시켜 https 설정이 무시되는 이유 때문이 아닌지 의심 됨
